fw/apm.c

```diff
@@ -950,18 +987,24 @@ tfw_apm_calc(TfwApmData *data)
 	rdidx = atomic_read(&data->stats.rdidx);
 	asent = &data->stats.asent[(rdidx + 1) % 2];
 
-	if (!tfw_apm_rbctl_update(data))
-		return;
-	tfw_apm_prnctl_calc(&data->rbuf, &data->rbctl, &pstats);
+	recalc = test_and_clear_bit(TFW_APM_DATA_F_RECALC, &data->flags);
+	if (!tfw_apm_rbctl_update(data, recalc))
+		return 0;
 
-	T_DBG3("%s: Percentile values may have changed.\n", __func__);
-	write_seqlock(&asent->seqlock);
-	memcpy_fast(asent->pstats.val, pstats.val,
-		    T_PSZ * sizeof(asent->pstats.val[0]));
-	atomic_inc(&data->stats.rdidx);
-	write_sequnlock(&asent->seqlock);
+	nfilled = tfw_apm_prnctl_calc(&data->rbuf, &data->rbctl, &pstats);
+	if (nfilled < T_PSZ) {
+		T_DBG3("%s: Percentile calculation incomplete.\n", __func__);
+		set_bit(TFW_APM_DATA_F_RECALC, &data->flags);
+	} else {
+		T_DBG3("%s: Percentile values may have changed.\n", __func__);
+		write_seqlock(&asent->seqlock);
+		memcpy_fast(asent->pstats.val, pstats.val,
+			    T_PSZ * sizeof(asent->pstats.val[0]));
+		atomic_inc(&data->stats.rdidx);
+		write_sequnlock(&asent->seqlock);
+	}
 
-	return;
+	return nfilled % T_PSZ;
```

Comment:
```suggestion
	return nfilled < T_PSZ;
```
and make the function to return `bool`? In my understanding this is no way to get `nfilled > T_PSZ`